cardano-api/src/Cardano/Api/Typed.hs

```diff
@@ -2853,6 +2855,75 @@ instance HasTextEnvelope (SigningKey StakePoolKey) where
     textEnvelopeType _ = "Node operator signing key"
     -- TODO: include the actual crypto algorithm name, to catch changes
 
+instance CastKeyRole StakePoolKey BlockIssuerKey where
+    castVerificationKey (StakePoolVerificationKey (Shelley.VKey vkey)) =
+      BlockIssuerVerificationKey (Shelley.VKey vkey)
+
+    castSigningKey (StakePoolSigningKey skey) =
+      BlockIssuerSigningKey skey
+
+--
+-- block issuer pool keys
+--
+
+data BlockIssuerKey
+
+instance HasTypeProxy BlockIssuerKey where
+    data AsType BlockIssuerKey = AsBlockIssuerKey
+    proxyToAsType _ = AsBlockIssuerKey
+
+instance Key BlockIssuerKey where
+
+    newtype VerificationKey BlockIssuerKey =
+        BlockIssuerVerificationKey (Shelley.VKey Shelley.BlockIssuer ShelleyCrypto)
+      deriving stock (Eq, Show)
+      deriving newtype (ToCBOR, FromCBOR)
+      deriving anyclass SerialiseAsCBOR
+
+    newtype SigningKey BlockIssuerKey =
+        BlockIssuerSigningKey (Shelley.SignKeyDSIGN ShelleyCrypto)
+      deriving stock (Show)
+      deriving newtype (ToCBOR, FromCBOR)
+      deriving anyclass SerialiseAsCBOR
+
+    deterministicSigningKey :: AsType BlockIssuerKey -> Crypto.Seed -> SigningKey BlockIssuerKey
+    deterministicSigningKey AsBlockIssuerKey seed =
+        BlockIssuerSigningKey (Crypto.genKeyDSIGN seed)
+
+    deterministicSigningKeySeedSize :: AsType BlockIssuerKey -> Word
+    deterministicSigningKeySeedSize AsBlockIssuerKey =
+        Crypto.seedSizeDSIGN proxy
+      where
+        proxy :: Proxy (Shelley.DSIGN ShelleyCrypto)
+        proxy = Proxy
+
+    getVerificationKey :: SigningKey BlockIssuerKey -> VerificationKey BlockIssuerKey
+    getVerificationKey (BlockIssuerSigningKey sk) =
+        BlockIssuerVerificationKey (Shelley.VKey (Crypto.deriveVerKeyDSIGN sk))
+
+    verificationKeyHash :: VerificationKey BlockIssuerKey -> Hash BlockIssuerKey
+    verificationKeyHash (BlockIssuerVerificationKey vkey) =
+        BlockIssuerKeyHash (Shelley.hashKey vkey)
+
+newtype instance Hash BlockIssuerKey =
+    BlockIssuerKeyHash (Shelley.KeyHash Shelley.BlockIssuer ShelleyCrypto)
+  deriving (Eq, Ord, Show)
+
+instance SerialiseAsRawBytes (Hash BlockIssuerKey) where
+    serialiseToRawBytes (BlockIssuerKeyHash (Shelley.KeyHash vkh)) =
+      Crypto.getHash vkh
+
+    deserialiseFromRawBytes (AsHash AsBlockIssuerKey) bs =
+      BlockIssuerKeyHash . Shelley.KeyHash <$> Crypto.hashFromBytes bs
+
+instance HasTextEnvelope (VerificationKey BlockIssuerKey) where
+    textEnvelopeType _ = "Node operator block issuer verification key"
+    -- TODO: include the actual crypto algorithm name, to catch changes
+
+instance HasTextEnvelope (SigningKey BlockIssuerKey) where
+    textEnvelopeType _ = "Node operator block issuer signing key"
+    -- TODO: include the actual crypto algorithm name, to catch changes
```

Comment:
Hmm. I don't think this is going to work.

The problem is we want to be able to read in either a genesis key delegate or a stake pool signing key, and then cast them to a block issuer to make the op cert. But those have different text envelope types right?

In fact I don't think we will actually need to serialise a block issuer key role at all.

Reply:
:+1: I will remove `BlockIssuerKey` just use `Shelley.coerceKeyRole` where appropriate